typing/ctype.ml

```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+     that have no free univars.
+   * In this first pass, [visited] is used as a mapping for previously visited
+     nodes, and must already contain all the free univars in [ty].
+   * The remaining (univar-closed) parts of the type are then instantiated
+     with [copy] using a common [copy_scope].
+   The reason to work in two passes lies in recursive types such as:
+     [let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m]
+   The type of [x#m] should be:
+     [ < n : 'c; p : < m : 'a. < n : 'a; p : 'b > > as 'b > ]
+   I.e., the universal type variable ['a] is both instantiated as a fresh
+   type variable ['c] when outside of its binder, and kept as universal
+   when under its binder.
+   Assumption: in the first call to [copy_sep], all the free univars should
+   be bound by the same [Tpoly] node. This guarantees that they are only
+   bound when under this [Tpoly] node, which has no free univars, and as
+   such is not part of the separate copy. In turn, this allows the separate
+   copy to keep the sharing of the original type without breaking its
+   binding structure.
+ *)
+let rec copy_sep ~copy_scope ~fixed ~free ~may_share
+    ~(visited : type_expr TypeHash.t) (ty : type_expr) =
   let univars = free ty in
   if is_Tvar ty || may_share && TypeSet.is_empty univars then
     if get_level ty <> generic_level then ty else
     let t = newstub ~scope:(get_scope ty) in
-    delayed_copy :=
-      lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty)))
-      :: !delayed_copy;
+    add_delayed_copy t copy_scope ty;
     t
   else try
-    let t, bound_t = List.assq (get_id ty) visited in
-    let dl = if is_Tunivar ty then [] else diff_list bound bound_t in
-    if dl <> [] && conflicts univars dl then raise Not_found;
-    t
+    TypeHash.find visited ty
   with Not_found -> begin
     let t = newstub ~scope:(get_scope ty) in
-    let desc = get_desc ty in
-    let visited =
-      match desc with
-        Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
-          (get_id ty, (t, bound)) :: visited
-      | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
-          visited
-      | Tlink _ | Tsubst _ ->
-          assert false
-    in
-    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~bound visited in
+    TypeHash.add visited ty t;
+    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~visited in
     let desc' =
-      match desc with
+      match get_desc ty with
       | Tvariant row ->
           let more = row_more row in
           (* We shall really check the level on the row variable *)
           let keep = is_Tvar more && get_level more <> generic_level in
+          (* In that case we should keep the original, but we still
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
           let more' = copy_rec ~may_share:false more in
           let fixed' = fixed && (is_Tvar more || is_Tunivar more) in
```

Comment:
In the `copy` version, the call to `copy_row` always use `fixed = true`, here the definition of `fixed'` is fairly different. Why? Relatedly, can you explain what the `fixed` parameter of `copy_sep` is meant to do? (I see that `instance_poly` and `instance_label` sometimes pass `true`, but I can't spot the logic there.)

Reply:
A polymorphic variant type whose row variable is a univar should be _fixed_, i.e. one cannot add or remove row fields through unification, and `Reither` can only contain one type. When we instantiate this univar into a normal one, the row is no longer fixed. The meaning of the `fixed` parameter of `copy_row` is to keep the fixedness information in the row. When it is false, the copy is no longer fixed.